### ISSUE TITLE
ParallelSSH governance and next steps

### ISSUE THREAD

**Issue by enkore** [ParallelSSH/parallel-ssh#386]
**The current situation**

- It unfortunately looks like @pkittenis is incommunicado since late 2022.
- I am an GitHub org member, but don't have admin access
  - Therefore nothing can be merged into `master` (a protected branch)
- Only Panos has permissions on the PyPI packages
  - No releases can be made to PyPI

**What has been done so far**

- I've tried contacting Panos on every email address I could find (seven different ones).
  - Unfortunately no response so far (hope he is alright)
- I squatted @parallel-ssh on GitHub in case we need to migrate to a new GH org
- I applied for a parallel-ssh org on PyPI which is pending as of this writing
- I created `next` branches in some of the repositories which can be used as a staging branch in the mean time

**Next steps**

- GitHub has no process for this if the owners of an org didn't make preparations
- PyPI has a process (PEP 541), which takes at least a few months. The upside here is that usually it's more or less "random" people trying to do this process, here we might have a bit of a headstart since I'm an org member

The main motivator for getting this unstuck is that the base packages (ssh-python and ssh2-python) need new releases to cope with Python 3.12 breakage.

CC @JacobCallahan @pletnes @pbrezina @McTonderski @ogajduse @Gauravtalreja1 @omkarkhatavkar

**Comment by JacobCallahan** [comment]
@enkore Thanks for leading this effort. It's unfortunate that we're in this position, but it does seem like we have some potential people willing to keep things going.

I think one thing that will definitely help with this in the future is documenting update processes and modernizing the build/release systems.

Luckily, Python's API is typically pretty stable and we're now at a 1 year Y-stream release cadence, so maintaining these packages shouldn't require a lot of time.

**Comment by enkore** [comment]
Pinging some downstream packages (according to github) to make them aware of this. Please excuse the noise if I included false positives in here.

https://github.com/persimmon-ai-labs/adept-inference @mtensor
https://github.com/marsupialtail/quokka @marsupialtail
https://github.com/BaguaSys/bagua @NOBLES5E
https://github.com/thelastpickle/cassandra-medusa @adejanovski @rzvoncek
https://github.com/scylladb/scylla-cluster-tests @fruch
https://github.com/nocproject/noc @aversant
https://github.com/carlmontanari/scrapli @carlmontanari

**Comment by fruch** [comment]
Thanks for the heads up

We are blocked on python 3.12 for our own libraries as well, so it would take a while till we'll be hitting this.

**Comment by enkore** [comment]
> PyPI has a process (PEP 541), which takes at least a few months. The upside here is that usually it's more or less "random" people trying to do this process, here we might have a bit of a headstart since I'm an org member

I've prepared an application for this and will submit it once the eligibility date (2023-12-03) passes (unless we hear back from Panos of course).

**Comment by rzvoncek** [comment]
Hello and thanks for reaching out! We're not blocked by this right now, but we're very happy you're looking into this nonetheless!

**Comment by pletnes** [comment]
We're running 3.10 instead of 3.11 because of this, but it's not really an issue for us, for now. But python versions EOL much faster these days so it's a good idea to get started.

**Comment by JacobCallahan** [comment]
@enkore I took a look through the issues you filed in this project and there are a number of things that need to be done to get this back in shape. I'm more than willing to jump in an take on a good amount of this, but coordination is a bit cumbersome via github alone.

If you'd like, we may be able to setup some kind of community video call with anyone interested in participating or just observing. If so, I can host a google meet.

**Comment by pkittenis** [comment]
I'll take a look at releasing updates packages soon.

**Comment by JacobCallahan** [comment]
@pkittenis Are you able to give us an update on the new releases? If this is something you need help with, we're here!

**Comment by McTonderski** [comment]
Hey @pkittenis Do you need any help with new release ? Let us know!

**Comment by JacobCallahan** [comment]
@enkore We may be in the same situation since there has been no activity.

**Comment by pbrezina** [comment]
@pkittenis  @enkore Bump. Is there any progress on the release?

**Comment by ulidtko** [comment]
@pkittenis ping

**Comment by ulidtko** [comment]
had better UX with `pip install fabric` https://docs.fabfile.org — has a CLI, and loads `~/.ssh/config` out of the box

**Comment by rzvoncek** [comment]
Hello all,

I've finally got to spend some time looking into this situation for our
project. We support Python up to 3.11 but we'd really like to add the newer
ones.

It seems that our main blocker is the parallel-ssh library. Even the forked
one seems to only go until Python 3.10, and even its most recent release
(2.12.0) still pulls in ssh-python and ssh2-python.
I can do some tricks
<https://python-poetry.org/docs/dependency-specification/#python-restricted-dependencies>
to
swap the ssh2-python for ssh2-python312 based on the Python version.
(However, I'm still not sure this actually works.
Also, I can't figure out any more complicated dependency management tricks
to handle the parallel-ssh situation.

So I'd like to ask for help. Has anybody faced something similar? Is there
any advice you could perhaps give me?

Thanks in advance,
Radovan

**Comment by ulidtko** [comment]
Meanwhile, in a similarly large community not so far away... :grin: @JacobCallahan @enkore 

![screenshot+meme](https://github.com/user-attachments/assets/b5652ba1-71f8-41a4-b642-9151e55a9ab6)

--------------------------

>  Is there any advice you could perhaps give me?

@rzvoncek from the PoV of library user like you, there aren't really "dependency management tricks" to work around a dead library. But since you asked for advice...

My personal choice is **boycotting Py3.12** (for several reasons, let's not get into that here), simply not updating to it yet — and I can continue doing so until [October next year](https://endoflife.date/python).

If you still want Py3.12, the choices I'd consider are:
* Migrate off the library. Find a replacement library, or write your own — then rewrite the code on your side that uses `parallel-ssh` functionality, to use something else. Then drop dependency on `parallel-ssh`. This thread shows why.
* Wade through the forks of `parallel-ssh` trying to find an API-compatible replacement, that is well-maintained. Then tell poetry to use that one instead of `parallel-ssh`; if such a replacement exists, that'll be the only change you'll need to make.

Another option would be to join @enkore in trying to revive this lib (then once PEP 541 completes, fix it for Py3.12, publish a release, upgrade to it) — but I'm not sure anyone can be helpful with this really, stuff like this is much easier to handle solo.

**Comment by pbrezina** [comment]
For the record, I added some patches to [pylibssh](https://github.com/ansible/pylibssh) and switch to it in pytest-mh, see the code: https://github.com/next-actions/pytest-mh/blob/master/pytest_mh/conn/ssh.py

**Comment by enkore** [comment]
fwiw I've been pulled from the work project that heavily relies on ssh-python and I doubt anyone else there is going to pick this up. I'm of course totally open to add/transfer ownership of that replacement org, if anyone is interested.

**Comment by pkittenis** [comment]
There are candidate new releases coming out.
